### PR TITLE
CF-874 specify tenants to pull from DB

### PR DESCRIPTION
**Pull request by VinnyQ**
* add functionality and test, map to new switch -t and --tenantIds

e.g.
```
sh /opt/cloudfeeds-ballista/bin/cloudfeeds-ballista.sh -t tenantid1
sh /opt/cloudfeeds-ballista/bin/cloudfeeds-ballista.sh --tenantIds tenantid1,tenantid2
sh /opt/cloudfeeds-ballista/bin/cloudfeeds-ballista.sh -runDate 2015-06-19 -t "tenantid1, tenantid2, tenantid'3"
```

### CODE REVIEW

**Comment by VinnyQ** [comment]
@ChandraAddala, @usnavi, @shintasmith please review when you can.

**Comment by usnavi** [comment]
One question:  What's with the single quotation mark in the tenant id in your example?

```
sh /opt/cloudfeeds-ballista/bin/cloudfeeds-ballista.sh -runDate 2015-06-19 -t "tenantid1, tenantid2, tenantid'3"
```

File: README.md [hunk 1]
```diff
@@ -54,6 +54,8 @@ The app expects the configuration files to be present at /etc/cloudfeeds-ballist
 
   -d <value> | --runDate <value>
          runDate is a date in the format yyyy-MM-dd. Data belonging to this date will be exported. Default value is yesterday.           
+  -t <value> | --tenantIds <value>
+        tenantIds is comma seprated list of tenant ids to be exported.  Default value is all data for all tenant ids.
```
Comment (shintasmith) [inline README.md]: It's funny... We have a story for Tiamat to change the behavior so that the default is not to process all tenant IDs. Right now, if you don't specify any option, Tiamat processes all tenant IDs. We were going to add an option that requires people to use --all-tenants option if they want to process all tenant IDs.

Do we need to make these two consistent? Just throwing a question. :-)

File: src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala [hunk 2]
```diff
@@ -22,5 +23,7 @@ class PreferencesDBQuery extends DBQuery {
     
   }
   
-
+  override def fetch(runDate: DateTime, tenantIds: Set[String], region: String, dataSource: DataSource, maxRowLimit: String): String = {
+    throw new NotImplementedException("not implemented for PreferencesDBQuery")
```
Comment (shintasmith) [inline src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala]: Just curious, why was this added?

File: README.md (hunk 1 above)
Comment (ChandraAddala) [inline README.md]: The requirement to add this functionality(to export by tenantid) to ballista came from testing. I dont really envision this being done in prod. I dont think we would benefit much by adding the flag(--all-tenants) to process all tenant ID's.

File: src/main/scala/com/rackspace/feeds/ballista/config/CommandOptionsParser.scala [hunk 3]
```diff
@@ -6,6 +6,7 @@ import scopt.Read
 
 case class CommandOptions(runDate: DateTime = DateTime.now.minusDays(1).withTimeAtStartOfDay(),
                           dbNames: Set[String] = AppConfig.export.from.dbs.dbConfigMap.keySet,
+                          tenantIds: Set[String] = null,
```
Comment (ChandraAddala) [inline src/main/scala/com/rackspace/feeds/ballista/config/CommandOptionsParser.scala]: Can you make this empty set instead of null?

File: src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala (hunk 2 above)
Comment (ChandraAddala) [inline src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala]: Is exporting preferences db by tenantId part of the requirement?

File: README.md (hunk 1 above)
Comment (shintasmith) [inline README.md]: Ok. Sounds good. :-)

File: src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala (hunk 2 above)
Comment (shintasmith) [inline src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala]: No, it's not a requirement. Now I see why it's added. Maybe it would be good to add a comment to note why we throw the exception (i.e: because exporting Prefs for a specific list of tenantIDs is not supported). May also change the exception to UnsupportedOperationException or something like that.

File: src/test/scala/com/rackspace/feeds/ballista/CommandProcessorITest.scala [hunk 4]
```diff
@@ -151,5 +151,23 @@ class CommandProcessorITest extends FunSuite with BeforeAndAfterAll {
...
+    assert(exitCode == 0, "Wrong exit code")
+
+    val remoteOutputLocation = AppConfig.export.from.dbs.dbConfigMap(testdbName)(DBProps.outputFileLocation)
+    val region = AppConfig.export.region
+    val remoteFilePath = s"$remoteOutputLocation/${region}_${testdbName}.gz"
+    val remoteSuccessFilePath = s"$remoteOutputLocation/${CommandProcessor.NON_DATA_FILE_PREFIX}$runDateStr/_SUCCESS"
 
+    assert(Files.exists(Paths.get(remoteFilePath)), s"Data for db[$testdbName] not present on remote server at path[$remoteFilePath]")
+    assert(Files.exists(Paths.get(remoteSuccessFilePath)), s"Success file not created on remote server at path[$remoteFilePath]")
+    assert(scala.io.Source.fromFile(remoteSuccessFilePath).mkString.replaceFirst("$\\n", "") == s"$testdbName=2", "Success file content mismatch")
```
Comment (ChandraAddala) [inline src/test/scala/com/rackspace/feeds/ballista/CommandProcessorITest.scala]: Is it possible to make an assertion that only data corresponding to those tenantId's got exported?

Comment (VinnyQ) [inline src/test/scala/com/rackspace/feeds/ballista/CommandProcessorITest.scala]: I am going to try, no promises.

File: src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala (hunk 2 above)
Comment (VinnyQ) [inline src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala]: Yeah, it's because the way the queries classes are setup, we have both the feeds query and the prefs query extend from a base trait.  We can split them out to 2 different traits (one specific for feed and one specific for prefs), or we can just make the prefs class not implement this specific method (prefs db does not have a correpsponding tenantid column for this query).  

I figured we go the easy and non-intrusive route, but if you guys want me to split out the trait to two separate traits (and also rewrite the command processor to know which query class to create), I can.

**Comment by VinnyQ** [comment]
@usnavi 
Re:  One question: What's with the single quotation mark in the tenant id in your example?

It's just there to test sql injection and to show the testers that it can be used in test cases.  Note that these are only samples in the PR description, not the official README or even a commit message.

File: src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala (hunk 2 above)
Comment (VinnyQ) [inline src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala]: I just realized that tenantid is actually mapped to the "id" field in the preferences table.  I will implement the query in this pull request also to allow user to pass in the --tenantid for prefereces to dump and scp.  Might as well do it now anyway, I am sure QEs will need it eventually.

Comment (VinnyQ) [inline src/main/scala/com/rackspace/feeds/ballista/queries/PreferencesDBQuery.scala]: code changed to implement tenantids search for preferencesdb also.

File: src/main/scala/com/rackspace/feeds/ballista/config/CommandOptionsParser.scala (hunk 3 above)
Comment (VinnyQ) [inline src/main/scala/com/rackspace/feeds/ballista/config/CommandOptionsParser.scala]: done.

File: src/test/scala/com/rackspace/feeds/ballista/CommandProcessorITest.scala (hunk 4 above)
Comment (VinnyQ) [inline src/test/scala/com/rackspace/feeds/ballista/CommandProcessorITest.scala]: Not going to do this for now, because it's pretty involved.  Gunziping the .gz, grabbing the content of the files, parsing the files, and matching the file data to the submitted data.

Note that you don't have assertions for the files content either for the default non-tenant cases from before, just row counts in the success file.

I did manually verify the files with the tenantids switch for specific tenantids (via all the steps above), and it is working as expected.

I think the QE team will have a better test framework setup to verify these archive exports.

File: src/main/scala/com/rackspace/feeds/ballista/queries/EntriesDBQuery.scala [hunk 5]
```diff
@@ -17,28 +17,40 @@ class EntriesDBQuery extends DBQuery {
...
+    logger.debug(s"Preparing query to extract data from partitioned entries table[$tableName] for runDate[$runDate], tenantIds[$tenantIds]")
     val runDateStr = dateTimeFormatter.print(runDate)
-    
+
     if (!isTableExist(tableName, dataSource)) {
       throw new RuntimeException("Partitioned table[$tableName] does not exist")
     }
+
+    var whereClause = ""
+    if (tenantIds != Set.empty) {
```
Comment (ChandraAddala) [inline src/main/scala/com/rackspace/feeds/ballista/queries/EntriesDBQuery.scala]: tenantids.nonEmpty call would do it.

File: src/main/scala/com/rackspace/feeds/ballista/service/DefaultExportSvc.scala [hunk 6]
```diff
@@ -81,8 +81,17 @@ class DefaultExportSvc(dbName: String) extends ExportSvc {
   def getQuery(queryParams: Map[String, Any]) = {
     val dbQuery = getInstance(dbConfigMap(dbName)(DBProps.queryClass))
     val region = AppConfig.export.region
-  
-    dbQuery.fetch(queryParams("runDate").asInstanceOf[DateTime], region, dataSource, AppConfig.export.maxRowLimit)
+    val tenantIds = queryParams("tenantIds").asInstanceOf[Set[String]]
+
+    if (tenantIds != Set.empty) {
```
Comment (ChandraAddala) [inline src/main/scala/com/rackspace/feeds/ballista/service/DefaultExportSvc.scala]: tenantIds.nonEmpty

**Comment by ChandraAddala** [comment]
lgtm;

**Comment by shintasmith** [comment]
lgtm2

File: src/main/scala/com/rackspace/feeds/ballista/queries/EntriesDBQuery.scala (hunk 5 above)
Comment (VinnyQ) [inline src/main/scala/com/rackspace/feeds/ballista/queries/EntriesDBQuery.scala]: done.

File: src/main/scala/com/rackspace/feeds/ballista/service/DefaultExportSvc.scala (hunk 6 above)
Comment (VinnyQ) [inline src/main/scala/com/rackspace/feeds/ballista/service/DefaultExportSvc.scala]: done.
